More detail on locks

KronosTheLate opened the pull request:
After reading the documentation on multi-threading, I did not catch the importance of locks and how to use them. This was made clear to me in [this](https://discourse.julialang.org/t/help-find-best-approach-to-streaming-data-on-raspberry-pi/105130/5) discourse thread.

In this PR, I have attempted to add the detail I was missing. It probably needs a second set of eyes to shorten the language and use a style that is more in line with the rest of the documentation.

KronosTheLate committed:
More detail on locks

After reading the documentation on multi-threading, I did not catch the importance of locks and how to use them. This was made clear to me in [this](https://discourse.julialang.org/t/help-find-best-approach-to-streaming-data-on-raspberry-pi/105130/5) discourse thread.

In this PR, I have attempted to add the detail I was missing. It probably needs a second set of eyes to shorten the language and use a style that is more in line with the rest of the documentation.

KronosTheLate committed:
minor wording change

vtjnash commented on doc/src/manual/multi-threading.md:
```diff
@@ -149,26 +149,37 @@ and nothing promised here can be assumed if you do not observe that
-               unlock(lk)
-           end
+           lock(my_lock)
+           my_operation()
```
there should always be a try/finally block following `lock`, or use one of the syntax forms (such as `@lock my_lock my_operations()`) that includes it for you

KronosTheLate replied:
Right, which is explained not in the immediate paragraph, but the one after. I find it to be a smoother conceptual journey to start with the simplest example, and build on it. Is the problem/worry that someone might copy that example without reading the context? If yes, would a good fix be to comment "#dont copy this code"?

We are splitting hairs, and I am fine with making all examples "propper". We can then annotate the `finally` clause with ”# ensures that lock is always unlocked", which also provides the very explicit guidance that I am trying to add here.

vtjnash replied:
I think most examples should be proper. The macro form might be best here, since it is simplest and also the most common form to use or encounter

KronosTheLate replied:
Alright, so instead of going basic and building up, we are showing the shortest and best first. This is worse for educational purposes, but better for those who want the TL;DR, which I guess is a lot of people when it comes to documentation. I will reorder the examples in the opposite order, and make them all propper.

KronosTheLate replied:
This should be resolved by the latest commit. take a look and see if you agree.

KronosTheLate committed:
Significant reordering and rewriting

I ended up significantly rewriting the section on locks, creating a seperate section for locks, reorder things, and add references between sections. I feel like the result is a document with better conceptual structure. 

I also rewrote a section on how to write a thread-safe with `@threads`, as I found the original sentence(s) extremely complicated and hard to read.

vtjnash committed:
Update multi-threading.md

vtjnash committed:
Update multi-threading.md
